### PR TITLE
FEATURE: [bybit] add trade info event

### PR DESCRIPTION
**Pull request by bailantaotao**

### CODE REVIEW

File: pkg/exchange/bybit/stream.go [hunk 1]
```diff
@@ -307,3 +330,70 @@ func (s *Stream) handleKLineEvent(klineEvent KLineEvent) {
...
+	for _, event := range events {
+		feeRate, found := s.symbolFeeDetails[event.Symbol]
+		if !found {
+			log.Warnf("unexpected symbol found, fee rate not supported, symbol: %s", event.Symbol)
+			continue
+		}
+
+		gTrade, err := event.toGlobalTrade(*feeRate)
+		if err != nil {
+			log.WithError(err).Errorf("unexpected trade: %+v", event)
```
Comment (c9s) [inline pkg/exchange/bybit/stream.go:344]: use "unable to convert trade: %+v" ?

File: pkg/exchange/bybit/stream.go [hunk 2]
```diff
@@ -307,3 +330,70 @@ func (s *Stream) handleKLineEvent(klineEvent KLineEvent) {
...
+type symbolFeeDetail struct {
+	bybitapi.FeeRate
+
+	BaseCoin  string
+	QuoteCoin string
+}
+
+// getFeeRate retrieves all fee rates from the Bybit API and then fetches markets to ensure the base coin and quote coin
+// are correct.
+func (e *Stream) getFeeRate(ctx context.Context) error {
```
Comment (c9s) [inline pkg/exchange/bybit/stream.go:360]: maybe call it getAllFeeRates() to let the method name document its behavior

File: pkg/exchange/bybit/types.go [hunk 3]
```diff
@@ -210,3 +212,153 @@ func (k *KLine) toGlobalKLine(symbol string) (types.KLine, error) {
...
+//   - FALSE
+//     IsMakerOrder = TRUE
+//     -> Side = Buy -> quote currency (USDT)
+//     -> Side = Sell -> base currency (BTC)
+//
+//     IsMakerOrder = FALSE
+//     -> Side = Buy -> base currency (BTC)
+//     -> Side = Sell -> quote currency (USDT)
+func (t *TradeEvent) CalculateFee(feeDetail symbolFeeDetail) (string, fixedpoint.Value) {
+	if feeDetail.MakerFeeRate > 0 {
```
Comment (c9s) [inline pkg/exchange/bybit/types.go:331]: use MakerFeeRate.Sign() > 0

File: pkg/exchange/bybit/types.go [hunk 4]
```diff
@@ -210,3 +212,153 @@ func (k *KLine) toGlobalKLine(symbol string) (types.KLine, error) {
...
+//
+//   - FALSE
+//     IsMakerOrder = TRUE
+//     -> Side = Buy -> quote currency (USDT)
+//     -> Side = Sell -> base currency (BTC)
+//
+//     IsMakerOrder = FALSE
+//     -> Side = Buy -> base currency (BTC)
+//     -> Side = Sell -> quote currency (USDT)
+func (t *TradeEvent) CalculateFee(feeDetail symbolFeeDetail) (string, fixedpoint.Value) {
```
Comment (c9s) [inline pkg/exchange/bybit/types.go:330]: Do you think we can use the feeModeFunctionNative function?

Comment (c9s) [reply pkg/exchange/bybit/types.go:330]: Oh seems like feeModeFunctionNative is for types.Order, not for trade

Comment (bailantaotao) [reply pkg/exchange/bybit/types.go:330]: That's a great idea! Perhaps we can extract price and quantity into interface functions. Then, both the order and trade can implement them. Ultimately, we can transition the order to use the aforementioned interface.

Comment (bailantaotao) [reply pkg/exchange/bybit/types.go:330]: We can do it in next PR

File: pkg/exchange/bybit/types.go [hunk 5]
```diff
@@ -210,3 +212,153 @@ func (k *KLine) toGlobalKLine(symbol string) (types.KLine, error) {
...
+	return feeDetail.TakerFeeRate.Mul(t.ExecQty)
+}
+
+func (t *TradeEvent) quoteCoinAsFee(feeDetail symbolFeeDetail) fixedpoint.Value {
+	baseFee := t.ExecPrice.Mul(t.ExecQty)
+	if t.IsMaker {
+		return feeDetail.MakerFeeRate.Mul(baseFee)
+	}
+	return feeDetail.TakerFeeRate.Mul(baseFee)
+}
```
[the remark this answers is not in the document]
Comment (bailantaotao) [reply pkg/exchange/bybit/types.go:357]: OK

File: pkg/exchange/bybit/stream.go [hunk 6]
```diff
@@ -27,35 +27,49 @@ var (
...
+
 //go:generate callbackgen -type Stream
 type Stream struct {
-	key, secret string
 	types.StandardStream
 
+	key, secret    string
+	marketProvider MarketInfoProvider
+	// TODO: update the fee rate at 7:00 am UTC; rotation required.
+	symbolFeeDetails map[string]*symbolFeeDetail
```
Comment (gx578007) [inline pkg/exchange/bybit/stream.go:43]: Will this be accessed from multiple goroutines?

**Review by gx578007 — APPROVED** [review]

File: pkg/exchange/bybit/stream.go (hunk 6 above)
Comment (bailantaotao) [reply pkg/exchange/bybit/stream.go:43]: Currently, data is only written during the initialization connection, so I haven't planned to handle it yet. When I get around to completing the TODOs, I will address it

**Review by c9s — APPROVED** [review]

**Pull request merged**